File: Java/IntelliJ/JavaLab/src/Placement/MaxSubArray.java
package Placement;

public class MaxSubArray
{

    private static int[] findMaxCrossingSubArray(int[] arr,int low,int mid,int high)
    {
        int leftSum = -99999;
        int leftIndex = 0;
        int sum = 0;

        for(int n=mid;n>=low;n--)
        {
            sum = sum + arr[n];

            if(sum>leftSum)
            {
                leftSum=sum;
                leftIndex = n;
            }
        }

        sum = 0;
        int rightSum = -99999;
        int rightIndex = 0;

        for(int n = mid+1;n<=high;n++)
        {
            sum=sum+arr[n];
            if(sum>rightSum)
            {
                rightSum=sum;
                rightIndex=n;
            }
        }

        return new int[] {leftIndex,rightIndex,leftSum+rightSum};
    }

    private static int[] findMaxSubArray(int[] arr,int low,int high)
    {
        final int SUM = 2;
        if(low==high)
        {
            return new int[] {low,high,arr[low]};
        }
        else
        {
            int mid = (low+high)/2;
            int[] left = findMaxSubArray(arr,low,mid);
            int[] right = findMaxSubArray(arr,mid+1,high);
            int[] cross = findMaxCrossingSubArray(arr,low,mid,high);

            if((left[SUM]>=right[SUM])&&(left[SUM]>=cross[SUM]))
            {
                return left;
            }
            else if((right[SUM]>=left[SUM])&&(right[SUM]>=cross[SUM]))
            {
                return right;
            }
            else
            {
                return cross;
            }

        }
    }

    public static void main(String[] args)
    {
        int[] arr = {-12,-6,34,-4,40,-9};

        int[] temp = findMaxSubArray(arr,0,5);

        System.out.print("Maximum sum is : "+temp[2]);
        System.out.print("\nStart index : "+temp[0]);
        System.out.print("\nEnd index : "+temp[1]);
    }
}

File: Python/MaxSubArray.py
#This program solves the maximum subarray problem

def find_max_crossing_subarray(a,low,mid,high):
    left_sum = -10000
    sum = 0
    max_left = mid
    for i in range(mid,low-1,-1):
        sum = sum+a[i]
        if(sum > left_sum):
            left_sum = sum
            max_left = i


    right_sum = -10000
    sum = 0
    max_right = mid+1
    for j in range(mid+1,high+1):
        sum = sum + a[j]
        if(sum > right_sum):
            right_sum = sum
            max_right = j

    return (max_left,max_right,left_sum+right_sum)


def find_max_subarray(a,low,high):
    if(low==high):
        return(low,high,a[low])
    else:
        mid = (low+high)//2
        (left_low,left_high,left_sum) = find_max_subarray(a,low,mid)
        (right_low,right_high,right_sum)=find_max_subarray(a,mid+1,high)
        (cross_low,cross_high,cross_sum)=find_max_crossing_subarray(a,low,mid,high)

        if((left_sum >= right_sum)and(left_sum >= cross_sum)):
            return (left_low,left_high,left_sum)
        elif((right_sum >= left_sum)and(right_sum >= cross_sum)):
            return(right_low,right_high,right_sum)
        else:
            return(cross_low,cross_high,cross_sum)


mylist = [8,-5,6,-5,-7,4,5,-3,9,9]
(low,high,sumsub)=find_max_subarray(mylist,0,9)

print("Low is : "+str(low))
print("\nHigh is : "+str(high))
print("\nSum is : "+str(sumsub))